syntax = "proto3";

package tetris_protocol;

enum BlockType {
	UNSPECIFIED = 0;
	I = 73;
	J = 74;
	L = 76;
	O = 79;
	S = 83;
	T = 84;
	Z = 90;
	EMPTY = 69;
	WALL = 87;
}

message PlayerName {
	string name = 1;
}

message Block {
	int32 start_column = 1;
	int32 lowest_start_row = 2;
	int32 rotations = 3;
	BlockType type = 4;
}

message Player {
	string device_guid = 1;
	string name = 2;
	int32 level = 3;
	int32 points = 4;
	bool ai = 5;
	Block current = 6;
	BlockType next = 7;
	int32 id = 8;
	int32 width = 9;
	int32 height = 10;
	int32 cleared_rows = 11;
	repeated BlockType board = 12;
}

message Game {
	int64 last_played_seconds = 1;
	repeated Player player = 2;
}

message HighScore {
	message Result {
		string name = 1;
		int32 points = 2;
		int32 rows = 3;
		int32 level = 4;
		int64 last_played_seconds = 5;
	}

	int32 columns = 1;
	int32 rows = 2;
	int32 level = 3;

	repeated Result results = 4;
}
